File: csv_detective/formats/siret.py
import re

proportion = 0.8
tags = ["fr"]
labels = [
    "siret",
    "siret d",
    "num siret",
    "siretacheteur",
    "n° siret",
    "coll siret",
    "epci",
]


def _is(val):
    """Détection des identifiants SIRET (SIRENE)"""
    if not isinstance(val, str):
        return False
    val = val.replace(" ", "")
    if not bool(re.match(r"^[0-9]{14}$", val)):
        return False

    # Vérification par clé de luhn du SIREN
    cle = 0
    pair = False
    for x in val[:9]:
        y = int(x) * (1 + pair)
        cle += y // 10 + y % 10
        pair = not pair
    if cle % 10 != 0:
        return cle % 10 == 0

    # Vérification par clé de luhn du SIRET
    cle = 0
    pair = len(val) % 2 == 0
    for x in val:
        y = int(x) * (1 + pair)
        cle += y // 10 + y % 10
        pair = not pair
    return cle % 10 == 0


_test_values = {
    True: ["13002526500013", "130 025 265 00013"],
    False: ["13002526500012"],
}

File: rust/src/formats/siret.rs
use super::Detector;
use crate::value::Value;
use super::siren::luhn_check;

pub struct SiretFormat;

impl SiretFormat {
    pub fn detect(&self, val: &str) -> Option<()> {
        let cleaned: String = val.chars().filter(|c| *c != ' ').collect();
        let bytes = cleaned.as_bytes();
        if bytes.len() != 14 || !bytes.iter().all(|b| b.is_ascii_digit()) {
            return None;
        }
        // SIREN check (first 9 digits)
        luhn_check(&bytes[..9])?;
        // SIRET check (all 14 digits)
        luhn_check(bytes)?;
        Some(())
    }
}

impl Detector for SiretFormat {
    fn name(&self) -> &'static str { "siret" }
    fn python_type(&self) -> &'static str { "string" }
    fn proportion(&self) -> f64 { 0.8 }
    fn mandatory_label(&self) -> bool { true }
    fn tags(&self) -> &'static [&'static str] { &["fr"] }
    fn labels(&self) -> &'static [(&'static str, f64)] {
        &[
            ("siret", 1.0), ("num siret", 1.0), ("siretacheteur", 1.0),
            ("n° siret", 1.0), ("coll siret", 1.0), ("epci", 1.0),
        ]
    }
    fn test(&self, val: &Value) -> bool { self.detect(val.raw()).is_some() }
}
